### PR TITLE
Migrate history folder into sdr - A new history.lua file and ReadHistory component

### PR DESCRIPTION
**Pull request by Hzj-jie**
This is the first step to migrate history into sdr, this issue has been discussed @ #1962 .

This change adds a new ReadHistory component to track and store read history. So components such as FileManagerHistory do not need to relay on history folder to build read history list.

### CODE REVIEW

File: frontend/docsettings.lua [hunk 1]
```diff
@@ -36,38 +36,38 @@ function DocSettings:purgeDocSettings(doc_path)
 end
 
 function DocSettings:open(docfile)
-    local history_path
-    local sidecar_path
+    -- TODO(zijiehe): Remove history_path, use only sidecar.
+    local new = {}
+    new.data = {}
```
Comment (houqp) [inline frontend/docsettings.lua]: local new = { data = {} }

File: frontend/apps/reader/readerui.lua [hunk 2]
```diff
@@ -329,6 +330,7 @@ end
 
 function ReaderUI:showReader(file)
     DEBUG("show reader ui")
+    ReadHistory:addItem(file)
```
Comment (houqp) [inline frontend/apps/reader/readerui.lua]: Is the history list bounded?

Comment (Hzj-jie) [inline frontend/apps/reader/readerui.lua]: Good point, currently no. Should we limit to 1000 items?

File: frontend/readhistory.lua [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+    end
+end
+
+function ReadHistory:init()
+    self:_read()
+    self:_readLegacyHistory()
+    self:_sort()
+end
+
+ReadHistory:init()
```
Comment (houqp) [inline frontend/readhistory.lua]: We should reduce side effect (io especially) at import time. Can we make the init method idempotent and call it from readerui and filemanager's init instead?

File: frontend/apps/reader/readerui.lua (hunk 2 above)
Comment (houqp) [inline frontend/apps/reader/readerui.lua]: I would even say go with 500 ;) I double anyone would want to keep track 500 books at once. But if you can test on the device and double check there is not a huge performance difference between 1000 and 500, then 1000 works for me too.

**Comment by houqp** [comment]
The rest looks good to me, thank you for cleaning this up :)

File: frontend/apps/reader/readerui.lua (hunk 2 above)
Comment (Frenzie) [inline frontend/apps/reader/readerui.lua]: I doubt it matters much on the machine end; I'd be worried more about whether the GUI allows for dealing efficiently with more than a page or three and whether it adds anything for the user. If I know it's a book I read sometime last year I use my well-organized folder structure or (rarely, if ever, even though I wrote it) the search functionality. That raises the question of whether search in history would be worthwhile, but again I'm inclined to think that'd add extra GUI complexity without really offering anything extra. To summarize, I think even as "little" as two or three hundred should suffice.

tl;dr I'm fine with 500, think 250 would suffice, and think 1000 wouldn't really profit anyone.

Comment (houqp) [inline frontend/apps/reader/readerui.lua]: Yeah, I actually prefer less number like 250. UI-wise, our menu widget does lazy rendering, so it handle very large list without issue. For me, I only use history menu for fast switching between the books that I am currently reading. Since I don't read many books at once, I never go pass the first page ;P

File: frontend/docsettings.lua (hunk 1 above)
Comment (Hzj-jie) [inline frontend/docsettings.lua]: Done.

File: frontend/apps/reader/readerui.lua (hunk 2 above)
Comment (Hzj-jie) [inline frontend/apps/reader/readerui.lua]: Since we are using O(n * log(n)) for sort, O(1) for delete, and O(n) for read. 1000 and 500 won't make much difference. I agree 500 books are out of track.
P.S. Until we finally removed history folder, we still need to read over 500 items, but only reduce after sort.

Comment (houqp) [inline frontend/apps/reader/readerui.lua]: Yeah, I am more interested in serialization cost and memory consumption for large tables. The current legacy history implementation is unfortunately unbouanded ;) Hopefully, we can get rid of it soon.

**Comment by houqp** [comment]
Can you rebase and squash those commits? It will make bisect easier in the future.

**Comment by houqp** [comment]
Cool, can you also fix the new errors in `make static-check`? The build will pass after you reduce the error count back to 55.

**Comment by Hzj-jie** [comment]
Oh, the failures have been fixed. :)

**Comment by houqp** [comment]
Thanks!

**Comment by Hzj-jie** [comment]
After this change, history list won't be updated if user moved or deleted or renamed files in history list. I doubt if it worth to track changed files.

**Comment by houqp** [comment]
Yeah, I don't think it's worth doing. They are not supposed to manually edit those meta data anyway unless something is wrong. We can reconsider it when someone actually complains to us ;)

**Comment by Frenzie** [comment]
@Hzj-jie Could you take a look at https://www.mobileread.com/forums/showthread.php?t=282824 ?
